[DataPipe] removing filter's inheritance from map

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64404
* #64398

This PR remove `filter`'s inheritance from `map`. This allows `filter` to not have a `__len__` function and that behavior is what we would like.

cc @VitalyFedyunin @ejguan

Differential Revision: [D30713120](https://our.internmc.facebook.com/intern/diff/D30713120)

NivekT committed:
removing filter's inheritance from map

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30713120).

ejguan commented on torch/utils/data/datapipes/iter/selecting.py:
```diff
@@ -1,13 +1,29 @@
+except ImportError:
+    DILL_AVAILABLE = False
+
+# Default function to return each item directly
+# In order to keep datapipe picklable, eliminates the usage
+# of python lambda function
+def default_filter_fn(data):
+    return True
```
Let's remove it here. We should expect user to provide a filter_fn

NivekT replied:
Done.

NivekT committed:
Update on "[DataPipe] removing filter's inheritance from map"


This PR remove `filter`'s inheritance from `map`. This allows `filter` to not have a `__len__` function and that behavior is what we would like.

cc @VitalyFedyunin @ejguan

Differential Revision: [D30713120](https://our.internmc.facebook.com/intern/diff/D30713120)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30713120).
